Extend concord errors

vogre opened the pull request:
Add names to errors for pretty-printing
Wrap all external errors with RequestError

Commit:
Extend concord errors

Add names to errors for pretty-printing
Wrap all external errors with RequestError

bergundy commented on package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "concord",
-  "version": "0.1.0",
+  "version": "0.2.0",
```
0.1.1?

vogre replied:
No preference really, since all interfaces need to be regenerated anyway.

bergundy commented on src/lib/common.ts:
```diff
@@ -4,13 +4,27 @@ import * as Ajv from 'ajv';
 export class ValidationError extends Error {
   constructor(message: string, public errors: any) {
     super(message);
+    this.name = 'ValidationError';
```
Or just:
```typescript
class ValidationError extends Error {
  public readonly name = 'ValidationError';
  ...
}
```

bergundy commented on src/lib/common.ts:
```diff
@@ -4,13 +4,27 @@ import * as Ajv from 'ajv';
 }
 
+export class RequestError extends Error {
+  public cause: any;
```
`name` can be hardcoded here.

vogre replied:
Cool trick, thanks!

bergundy commented on src/lib/common.ts:
```diff
@@ -4,13 +4,27 @@ import * as Ajv from 'ajv';
+  public cause: any;
+  public method: string;
+  public options: any;
+  constructor(message: string, cause: any, method: string, options: any) {
```
Why not initialise these in the params: `public readonly cause: any`?

bergundy commented on src/lib/common.ts:
```diff
@@ -4,13 +4,27 @@ import * as Ajv from 'ajv';
 }
 
+export class RequestError extends Error {
+  public cause: any;
```
Prefer readonly attributes.

bergundy commented on src/test/test_rpc.ts:
```diff
@@ -766,7 +766,17 @@ await new Promise((resolve, reject) => {
 server.close();
-await expect(client.bar('heh')).to.eventually.be.rejectedWith(/connect ECONNREFUSED/);
+try {
+  await client.bar('heh');
```
You can use `.to.eventually.be.rejectedWith`

vogre replied:
I can't test the error object this way.

bergundy replied:
You can, just `const err = await expect(client.bar('heh')).to.eventually.be.rejected`.
Please do, adding `expect(false)` is more error prone, too easy to miss.

vogre replied:
Thanks!

It was not obvious from chai documentation...

bergundy commented on src/test/test_rpc.ts:
```diff
@@ -814,7 +824,17 @@ await new Promise((resolve, reject) => {
 const { address, port } = (server.address() as AddressInfo);
 const client = new TestClient('http://' + address + ':' + port);
-await expect(client.bar('heh')).to.eventually.be.rejectedWith(Error, '500 - Internal Server Error');
+try {
```
See above.
If `rejectedWith` is not good enough you can use:
```
const err = await client.bar('heh').catch((err) => err);
```

vogre replied:
This doesn't really work if there's no error.

bergundy commented on src/test/test_rpc.ts:
```diff
@@ -871,7 +899,7 @@ import fetch from 'node-fetch';
 
 export default async function test(client: TestClient) {
-  const url = (client as any /* access protected member */).serverUrl;
+  const url = client.serverUrl;
```
That makes sense..

bergundy commented on templates/ts/client-browser.ts:
```diff
@@ -80,40 +80,48 @@ export class {{name}}Client {
+        responseBody = await response.json();
+      }
+    } catch (err) {
+      throw new RequestError(err.message, err, "{{name}}", { serverUrl: this.serverUrl, ...this.options, ...options });
```
Can we use single quotes for `"{{name}}"`?

bergundy commented on templates/ts/client-browser.ts:
```diff
@@ -80,40 +80,48 @@ export class {{name}}Client {
+        responseBody = await response.json();
+      }
+    } catch (err) {
+      throw new RequestError(err.message, err, "{{name}}", { serverUrl: this.serverUrl, ...this.options, ...options });
```
You can use mergedOptions

vogre replied:
Most likely will have a signal object which would look bad in a log line.

bergundy commented on templates/ts/client-browser.ts:
```diff
@@ -80,40 +80,48 @@ export class {{name}}Client {
+        method: 'POST',
+      });
+      isJSON = (response.headers.get('content-type') || '').startsWith('application/json');
+      if (isJSON && (response.status >= 200 && response.status < 300 || response.status === 400 || response.status === 500)) {
```
Is it necessary to check the response status code?

vogre replied:
Added comment.

bergundy replied:
I'm not sure we should limit parsing the response to these error codes, it makes this line too dependant on the other conditions and likely to break.

bergundy commented on templates/ts/client-browser.ts:
```diff
@@ -80,40 +80,48 @@ export class {{name}}Client {
     if (response.status >= 200 && response.status < 300) {
       const validator = this.validators.{{{name}}};
-      const wrapped = { returns: isJSON ? await response.json() : undefined }; // wrapped for coersion
+      const wrapped = { returns: isJSON ? responseBody : undefined }; // wrapped for coersion
```
No need to check if `isJSON` anymore

bergundy commented on templates/ts/client-browser.ts:
```diff
@@ -80,40 +80,48 @@ export class {{name}}Client {
-      if (body.name === 'ValidationError') {
-        throw new ValidationError(body.message, body.errors);
+      if (responseBody.name === 'ValidationError') {
+        throw new ValidationError(responseBody.message, responseBody.errors);
```
Should we double check that responseBody is an object and has `message` and `errors`?

vogre replied:
In general we can probably JSON schema check the error responses, but I don't think it's likely we'll see a problem with them.

bergundy replied:
Agree

bergundy commented on templates/ts/client-node.ts:
```diff
@@ -1,8 +1,8 @@
 // tslint:disable
 import fetch from 'node-fetch';
-import { RequestInit } from 'node-fetch';
+import { RequestInit, Response } from 'node-fetch';
```
Not reviewing this file, assuming same diff as the browser code.

vogre replied:
Verified by diff.

bergundy commented on templates/ts/interfaces.ts:
```diff
@@ -2,10 +2,18 @@
 export const schema = {{{schema}}};
 
 export class InternalServerError extends Error {
+  constructor(message: string) {
```
I'd put the name as `public readonly name = 'InternalServerError'`.

bergundy reviewed (commented):
Overall LGTM.
nit: please use single quotes for all strings.

Commit:
CR: Use TypeScript syntax for constructors

Fix singlequotes
Add comment about status code checks
Remove isJSON checks
Remove variable name from error message

vogre reviewed (commented):
Thanks!

bergundy commented on src/lib/common.ts:
```diff
@@ -11,6 +12,18 @@ export interface ClassValidator {
+  public readonly name = 'RequestError';
+  constructor(
+    message: string,
+    public readonly cause: any,
```
Could you please document what cause is?

bergundy reviewed (commented):
Thanks!

Commit:
CR: Restore using chai for rejection assertions

Remove status codes logic and always consume responses
Document RequestError.cause

bergundy commented on templates/ts/client-browser.ts:
```diff
@@ -80,40 +80,54 @@ export class {{name}}Client {
+        responseText = await response.text();
+      }
+    } catch (err) {
+      throw new RequestError(err.message, err, '{{name}}', { serverUrl: this.serverUrl, ...this.options, ...options });
```
Maybe use `mergedOptions`?

bergundy reviewed (approved):
Thanks!

Commit:
CR: Use mergedOptions in errors

Unify timeout code between client and server to reduce diffs

vogre commented:
Thanks!

PTAL, I brought timeout code to browser generated code, it seems to be passing tests.

bergundy reviewed (approved):
Thanks!
